### PR TITLE
Improving tests

### PR DESCRIPTION
**Pull request by Reallife**
| Q             | A
| ------------- | ---
| Bug fix?      | No
| New feature?  | Yes

### Added

- `PrometheusValuesDictionary`

### Changed

- Improving labels and types formatting for `PrometheusFormatter`
- Improving tests of `PrometheusFormatter`

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I wrote unit tests for my code

### CODE REVIEW

File: src/Formatters/Dictionaries/PrometheusValuesDictionary.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
+<?php
+
+declare(strict_types=1);
+
+namespace AvtoDev\AppMetrics\Formatters\Dictionaries;
```
Comment (Reallife) [inline src/Formatters/Dictionaries/PrometheusValuesDictionary.php:5]: May be move outside of the `Formatters` directory?

File: src/Formatters/Dictionaries/PrometheusValuesDictionary.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    public const NEGATIVE_INF = '-Inf';
+
+    /**
+     * Returns all dictionary entries.
+     *
+     * @return array
+     */
+    public static function all(): array
+    {
+//        return (new \ReflectionClass(static::class))->getConstants();
```
Comment (Reallife) [inline src/Formatters/Dictionaries/PrometheusValuesDictionary.php]: May be better to get constants using reflection?

File: tests/Formatters/PrometheusFormatterTest.php [hunk 3]
```diff
@@ -135,8 +134,133 @@ public function value()
...
+        foreach ($data_sets as [$input, $expected]) {
+            $mock = $mock
+                ->shouldReceive('labels')
+                ->once()
+                ->andReturn($input)
+                ->getMock();
+
+            $result = $this->formatter->format([$mock]);
+
+            $this->assertSame("foo{{$expected}} 1", $result);
```
Comment (Reallife) [inline tests/Formatters/PrometheusFormatterTest.php]: Need to improve label formatting. Now we can get `foo{} 1`. That's wrong.

Comment (Reallife) [reply tests/Formatters/PrometheusFormatterTest.php]: fixed

File: src/Formatters/Dictionaries/PrometheusValuesDictionary.php (hunk 2 above)
Comment (Reallife) [reply src/Formatters/Dictionaries/PrometheusValuesDictionary.php]: ~~May be `public const ALL = self;`~~

File: src/Formatters/PrometheusFormatter.php [hunk 4]
```diff
@@ -73,53 +74,130 @@ public function format(iterable $metrics): string
...
-            : (string) $value;
+        if (\is_int($value) || \is_float($value)) {
+            return (string) $value;
+        }
+
+        if (\is_bool($value)) {
+            return $value ? '1' : '0';
+        }
+
+        if (\is_string($value) && \in_array($value, PrometheusValuesDictionary::all(), true)) {
```
Comment (Reallife) [inline src/Formatters/PrometheusFormatter.php]: case sensitive?

File: src/Formatters/PrometheusFormatter.php [hunk 5]
```diff
@@ -73,53 +74,130 @@ public function format(iterable $metrics): string
...
+            }
+            // https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels
+            if (! preg_match('/^[a-zA-Z_][a-zA-Z0-9_]*$/', (string) $key)) {
+                continue;
+            }
+
+            $formatted_value = $value;
+
+            if (\is_bool($value)) {
+                $formatted_value = $value ? 'true' : 'false';
```
Comment (Reallife) [inline src/Formatters/PrometheusFormatter.php:192]: 1/0 or true/false. Ask Ivan

File: tests/Formatters/PrometheusFormatterTest.php [hunk 6]
```diff
@@ -135,8 +134,173 @@ public function value()
...
+            [['foo' => 123], '{foo="123"}'],
+            [['foo' => 12.3], '{foo="12.3"}'],
+            [['foo'], ''],
+            [['foo' => \tmpfile()], ''],
+            [['foo' => function () {
+            }], ''],
+            [['foo' => []], ''],
+        ];
+
+        $mock = m::mock(\implode(', ', [MetricInterface::class, HasLabelsInterface::class]))
```
Comment (Reallife) [inline tests/Formatters/PrometheusFormatterTest.php]: Use anonymous class witch changing methods privacy.

File: src/Formatters/Dictionaries/PrometheusValuesDictionary.php (hunk 2 above)
Comment (Reallife) [reply src/Formatters/Dictionaries/PrometheusValuesDictionary.php]: Leave a manual filling constants array. If developer forgot to add new constant to `all()` method, tests must fails.

File: src/Formatters/Dictionaries/PrometheusValuesDictionary.php (hunk 1 above)
Comment (Reallife) [reply src/Formatters/Dictionaries/PrometheusValuesDictionary.php:5]: Leave as is

File: src/Formatters/PrometheusFormatter.php (hunk 5 above)
Comment (Reallife) [reply src/Formatters/PrometheusFormatter.php:192]: Asked. Ivan says, that better to use "true"/"false"

File: src/Formatters/PrometheusFormatter.php (hunk 4 above)
Comment (Reallife) [reply src/Formatters/PrometheusFormatter.php]: fixed

File: tests/Formatters/PrometheusFormatterTest.php (hunk 6 above)
Comment (Reallife) [reply tests/Formatters/PrometheusFormatterTest.php]: Used configurable by passed parameters Mockery's mock

**Comment by tarampampam** [comment]
@Reallife Formatters list with formatting rules is required in readme file. Code looks fine!

File: README.md [hunk 7]
```diff
@@ -53,6 +53,32 @@ some_metric{foo="bar"} 1
...
 Secret verification can be completed using sending get-parameter `?secret=XXX` or HTTP header `X-SECRET`.
 
+### Formatters
+
+Out of the box this package provides `json` and `prometheus` metrics formats. Feel free to write your own formatters.
+
+#### Prometheus formatter
+
+Prometheus formatter uses some constants for metric's values(`Nan`, `+Inf`, `-Inf`) and types(`COUNTER`,`SUMMARY`,`GAUGE`,`HISTOGRAM`, `UNTYPED`). You can find it in
+[`\AvtoDev\AppMetrics\Formatters\Dictionaries\PrometheusValuesDictionary`](https://github.com/avto-dev/app-metrics-laravel/blob/master/src/Formatters/Dictionaries/PrometheusValuesDictionary.php) and
```
Comment (tarampampam) [inline README.md]:
```
[`\AvtoDev\AppMetrics\Formatters\Dictionaries\PrometheusValuesDictionary`](https://github.com/avto-dev/app-metrics-laravel/blob/master/src/Formatters/Dictionaries/PrometheusValuesDictionary.php)
```
->
```
[`PrometheusValuesDictionary`](./src/Formatters/Dictionaries/PrometheusValuesDictionary.php)
```

> in all places

File: README.md [hunk 8]
```diff
@@ -53,6 +53,32 @@ some_metric{foo="bar"} 1
...
+
+Prometheus formatter uses some constants for metric's values(`Nan`, `+Inf`, `-Inf`) and types(`COUNTER`,`SUMMARY`,`GAUGE`,`HISTOGRAM`, `UNTYPED`). You can find it in
+[`\AvtoDev\AppMetrics\Formatters\Dictionaries\PrometheusValuesDictionary`](https://github.com/avto-dev/app-metrics-laravel/blob/master/src/Formatters/Dictionaries/PrometheusValuesDictionary.php) and 
+[`\AvtoDev\AppMetrics\Metrics\HasTypeInterface`](https://github.com/avto-dev/app-metrics-laravel/blob/master/src/Metrics/HasTypeInterface.php)
+
+Keep in mind that `Prometheus` accepts only numeric metrics values. This formatter will try to cast metrics values to numeric format (`integer` or `float`). And will set `Nan` if casting to numeric is impossible. Examples for formatting values:
+
+| Input | Output |
+| ----- | ------ |
+| 1.2 | '1.2' |
```
Comment (tarampampam) [inline README.md]:
```
'1.2'
```
->
```
`1.2`
```

> in all places

File: tests/Formatters/Dictionaries/PrometheusValuesDictionaryTest.php [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+declare(strict_types=1);
+
+namespace AvtoDev\AppMetrics\Tests\Formatters\Dictionaries;
+
+use AvtoDev\AppMetrics\Tests\AbstractUnitTestCase;
+use AvtoDev\AppMetrics\Formatters\Dictionaries\PrometheusValuesDictionary;
+
+/**
+ * Class PrometheusValuesDictionaryTest.
```
Comment (tarampampam) [inline tests/Formatters/Dictionaries/PrometheusValuesDictionaryTest.php]: del

File: tests/Formatters/Dictionaries/PrometheusValuesDictionaryTest.php [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    public function testAllMethod(): void
+    {
+        $constants = (new \ReflectionClass(PrometheusValuesDictionary::class))->getConstants();
+        $this->assertEmpty(\array_diff($constants, PrometheusValuesDictionary::all()));
+    }
+
+    /**
+     * @return array
+     */
+    protected function getExpectedEntries():array
```
Comment (tarampampam) [inline tests/Formatters/Dictionaries/PrometheusValuesDictionaryTest.php]: This method is used in a one test. So, this method is required?

File: tests/Formatters/PrometheusFormatterTest.php [hunk 11]
```diff
@@ -86,57 +87,210 @@ public function testFormatWithPassingTwoMetric(): void
...
+            'baz' => 'yahoo',
+        ];
+
+        $mock = $this->getMetricMock('blah', true, 'fake_type', $labels, 'fake');
+
+        $result = $this->formatter->format([$mock]);
+
+        $this->assertRegExp("~HELP blah fake\n~", $result);
+        $this->assertRegExp("~TYPE blah UNTYPED\n~", $result);
+        $this->assertRegExp('~blah{foo="1",bar="3.14",baz="yahoo"} 1~', $result);
```
Comment (tarampampam) [inline tests/Formatters/PrometheusFormatterTest.php]: regexp or `->assertSame`?

File: README.md [hunk 12]
```diff
@@ -53,6 +53,32 @@ some_metric{foo="bar"} 1
 
 Secret verification can be completed using sending get-parameter `?secret=XXX` or HTTP header `X-SECRET`.
 
+### Formatters
+
+Out of the box this package provides `json` and `prometheus` metrics formats. Feel free to write your own formatters.
+
+#### Prometheus formatter
+
+Prometheus formatter uses some constants for metric's values(`Nan`, `+Inf`, `-Inf`) and types(`COUNTER`,`SUMMARY`,`GAUGE`,`HISTOGRAM`, `UNTYPED`). You can find it in
```
Comment (tarampampam) [inline README.md]: In a one line, please (without line breaks)

**Review by tarampampam — CHANGES_REQUESTED** [review]

File: README.md (hunk 7 above)
Comment (Reallife) [reply README.md]: Fixed

File: README.md (hunk 8 above)
Comment (Reallife) [reply README.md]: Fixed

File: tests/Formatters/Dictionaries/PrometheusValuesDictionaryTest.php (hunk 9 above)
Comment (Reallife) [reply tests/Formatters/Dictionaries/PrometheusValuesDictionaryTest.php]: Fixed

File: tests/Formatters/Dictionaries/PrometheusValuesDictionaryTest.php (hunk 10 above)
Comment (Reallife) [reply tests/Formatters/Dictionaries/PrometheusValuesDictionaryTest.php]: Removed

File: tests/Formatters/PrometheusFormatterTest.php (hunk 11 above)
Comment (Reallife) [reply tests/Formatters/PrometheusFormatterTest.php]: Fixed

File: README.md (hunk 12 above)
Comment (Reallife) [reply README.md]: Fixed

**Review by tarampampam — APPROVED** [review]

**Pull request merged**